### ISSUE TITLE
Use x-ratelimit-reset-after 

### ISSUE THREAD

**Issue by Mehgugs** [SinisterRectus/Discordia#211]
After a discussion in #api it became clear to me that `x-ratelimit-reset-after` would be better to use in the rate limiter because it ensures system clock differences cannot cause a limit to be hit. It should be pretty easy to add this change + milisecond precision.

**Comment by SinisterRectus** [comment]
Can you link the discussion here?

**Comment by Mehgugs** [comment]
Discussion from #api: 
<https://discordapp.com/channels/81384788765712384/381887113391505410/627202742007365632>
<https://discordapp.com/channels/81384788765712384/381887113391505410/627205322288922645>

**Comment by Mehgugs** [comment]
Essentially discord's reset-after header might be a bit slower by some ms but it has stronger guarentees. It's better to be a little slower than to potentially unlock a route prematurely.

**Comment by SinisterRectus** [comment]
I haven't rushed to implement this because as far as I know, it only applies to reaction ratelimiting. The discussion about clock inaccuracies does not apply to Discordia.

**Comment by Mehgugs** [comment]
I think it does apply to discordia

**Comment by SinisterRectus** [comment]
Can you elaborate on this? `os.time` isn't used for ratelimits.

**Comment by Mehgugs** [comment]
```lua
function Date.parseHeader(str)
	local day, month, year, hour, min, sec = str:match(
		'%a+, (%d+) (%a+) (%d+) (%d+):(%d+):(%d+) GMT'
	)
	return Date.parseTableUTC {
		day = day, month = months[month], year = year,
		hour = hour, min = min, sec = sec, isdst = false,
	}
end
```

```lua
function Date.parseTableUTC(tbl)
	return time(tbl) + offset()
end
```

```lua
local function offset() -- difference between *t and !*t
	return difftime(time(), time(date('!*t')))
end
```

**Comment by SinisterRectus** [comment]
Oh that. Yeah that's not dependent on the system clock. It's using the relative difference between two os.time calls to correct the error on one os.time call, whether that error is expected (timezone differences) or not (an inaccurate system clock). The only chance of this failing is if your system clock changed between the call to time(tbl) and offset() during addition.

The inaccuracy case discussed in #api is when you use your system clock to determine Discord's start time for the ratelimt. Discordia does not so this; it uses the Date header as the start time.

**Comment by SinisterRectus** [comment]
If you can give me a test case where parseHeader or parseTableUTC has unexpected results, then that is a different issue that would need fixing.

**Comment by SinisterRectus** [comment]
Now accepting PRs:

- Use `X-RateLimit-Reset-After` to determine how long to wait before unlocking a route mutex when requests are depleted.
- Implement proper `X-RateLimit-Precision` usage.

A robust solution should probably confirm that the value for `X-RateLimit-Reset-After` is numeric, especially if the precision header is set. Discordia does not currently do this.

**Comment by SinisterRectus** [comment]
Implemented with 0a4949fb9f0f054716c6c6c9037eb2b093cf9ef1. The most notable difference here will be an increase in the speed at which sequential reactions are added or removed.
